### DISCUSSION TITLE
pool always empty when using attr-sql plugin

### DISCUSSION THREAD

**Discussion by RobertoTired** [strongswan/strongswan#2331] [Q&A] [answered]
Hello all

I am currently using strongswan on my openwrt router : strongSwan 5.9.11, Linux 5.15.150, mips
I already have a perfectly fine working config but I need to update it to assign static virtual IP to my clients based on their identity.
I have added attr-sql in the list of plugin : 

attr-sql {

    # Release all online leases during startup.  Disable this to share the DB
    # between multiple VPN gateways.
    # crash_recovery = yes

    # Database URI for attr-sql plugin used by charon. If it contains a
    # password, make sure to adjust the permissions of the config file
    # accordingly.
    database = sqlite:///etc/config/pool.db

    # Enable logging of SQL IP pool leases.
    # lease_history = yes

    # Whether to load the plugin. Can also be an integer to increase the
    # priority of this plugin.
    load = 2 # <= put 2 here but same behaviour with yes

}

I have initialized sqlite db using this script : https://github.com/strongswan/strongswan/blob/master/src/pool/sqlite.sql

Plugin seems to start just fine : 
Fri Jul  5 13:24:02 2024 daemon.info ipsec: 00[LIB] loaded plugins: charon attr-sql pkcs11 aes des blowfish rc2 sha2 sha1 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp pem openssl pkcs8 af-alg gmp gmpdh curve25519 chapoly xcbc cmac hmac kdf ctr ccm gcm drbg bliss sqlite kernel-netlink socket-default socket-dynamic stroke vici smp updown xauth-generic duplicheck

As first test I just removed the static pool in config file and just let the link in connection :

in swanctl.conf

connections {
  all_remote_ikev2 {
    local_addrs = %any
    remote_addrs = %any
    fragmentation = yes
    pools = ipv4pool
[...]
#pools {
#  ipv4pool {
#    addrs = 192.168.10.0/24 <= I commented this part

Created the pool in DB like this : 
ipsec pool --add ipv4pool --start 192.168.10.1 --end 192.168.10.10

Output of ipsec pool is strange : 

ipsec pool --status
no dns servers found.
no nbns servers found.
    name           start             end  timeout   size      online       usage
**invalid result type supplied**
ipv4pool    192.168.10.1   192.168.10.10   static     10     0 ( 0%)     0 ( 0%) 

When I restart daemon no pool is found :

Fri Jul  5 13:24:05 2024 daemon.info ipsec: 00[DMN] load-all: no pools found, 0 unloaded

swanctl -A display nothing.

Maybe there is something I didn't understand from documentation...

Thank you for your help !

**Comment by tobiasbrunner** [comment, marked as answer]
> Fri Jul 5 13:24:05 2024 daemon.info ipsec: 00[DMN] load-all: no pools found, 0 unloaded

These are only the pools loaded by the load-pool vici command. So this output is expected if you don't define any pools in swanctl.conf.

> swanctl -A display nothing.

That command also only lists pools defined in swanctl.conf.

The SQL-based pools you have to query using `ipsec pool`.

> **invalid result type supplied**

I can't reproduce this. Did you modify the sources somehow? It is logged if an unknown data type is passed for a column when querying data. Since it's only logged once, I also wonder which SQL query causes it exactly (there is the main query and three sub-queries per pool). Maybe it's some platform issue (or with a specific libsqlite3 version).

**Comment by RobertoTired** [comment]
Ok thanks for your answer.
I use a home compiled version of openwrt on a edge router, so yes the issue can be related to this. No modification of source, I just use the plain make menuconfig to configure the compilation process. It's the first time I see a specific issue due to this config.
SQLite version is 3.41.2.
I will try to debug more and if I find something usefull I will come back here.
